src/banded/BandedMatrix.jl

```diff
@@ -87,9 +89,8 @@ end
 
 Base.copy(B::BandedMatrix) = _BandedMatrix(copy(B.data), B.m, B.l, B.u)
 
-# Base.similar(a::BandedMatrix, ::Type{T}, dims::Dims{2}) where {T}    = BandedMatrix{T}(undef, dims, bandwidths(a))
-
-Base.promote_rule(::Type{BandedMatrix{T}}, ::Type{BandedMatrix{V}}) where {T,V} = BandedMatrix{promote_type(T,V)}
+Base.promote_rule(::Type{BandedMatrix{T, Matrix{T}}}, ::Type{BandedMatrix{V, Matrix{V}}}) where {T,V} =
```

Comment:
Shouldn't this just be:
```julia
Base.promote_rule(::Type{BandedMatrix{T, M}}, ::Type{BandedMatrix{V, N}}) where {T,V,N,M} =
BandedMatrix{promote_type(T,V), promote_type(N,M)}

Reply:
It's not so simple for non-scalar types. Indeed:

```
julia> promote_rule(Matrix{Float64}, SparseMatrixCSC{Int64})
Union{}
```

Nominally, here you could say the result should be `Matrix{Float64}`. But what if you switched the two element types? and what if it were not a sparse matrix but a gpu matrix?

The meaning of promote_rule does not extend well to containers.

Reply:
Never call `promote_rule`, you should call `promote_type`:
```julia
julia> promote_type(Matrix{Float64}, SparseMatrixCSC{Int64})
AbstractArray{T,2} where T
```
So I think my proposal above is reasonable, in that 
```julia
promote_type(BandedMatrix{T,Matrix{T}}, BandedMatrix{T,SparseMatrixCSC{T,Int}}) 
```
will return
```julia
BandedMatrix{T,AbstractMatrix{T}}
```